File: fhir-bridge/dotnet/src/FhirBridge/PedigreeToFhir.cs
using System.Text;
using System.Text.Json;

namespace FhirBridge;

public sealed class MappingException : Exception
{
    public MappingException(string message) : base(message) { }
}

public sealed record PedigreeMappingResult(
    string BundleJson,
    string ProbandReference,
    int MappedResourceCount,
    IReadOnlyList<string> Warnings);

public static class PedigreeToFhir
{
    private const string AdminGenderSystem = "http://hl7.org/fhir/administrative-gender";

    public static PedigreeMappingResult ToFhirBundle(PedigreeDetail detail)
    {
        var relations = ProbandRelations.FromProband(detail)
            ?? throw new MappingException(
                $"Pedigree {detail.Id} has no proband: set one before pushing to FHIR.");

        var probandReference = $"Patient/{relations.Proband.Id}";
        var warnings = new List<string>();
        var entries = new List<Action<Utf8JsonWriter>>();

        foreach (var view in relations.Relatives)
        {
            var coding = RelationMap.ToFhir(view.RelativeType);
            if (coding is null)
            {
                warnings.Add(
                    $"skipped {view.Individual.DisplayName}: no FHIR code for relative_type '{RelativeTypeWire.Of(view.RelativeType)}'.");
                continue;
            }
            entries.Add(writer => WriteEntry(writer, view.Individual, probandReference, coding));
        }
        foreach (var unlabelled in relations.Unlabelled)
        {
            warnings.Add(
                $"skipped {unlabelled.DisplayName}: could not derive a supported relation to the proband.");
        }

        var bundleJson = WriteBundle(entries);
        return new PedigreeMappingResult(bundleJson, probandReference, entries.Count, warnings);
    }

    private static string WriteBundle(IReadOnlyList<Action<Utf8JsonWriter>> entryWriters)
    {
        using var stream = new MemoryStream();
        using (var writer = new Utf8JsonWriter(stream))
        {
            writer.WriteStartObject();
            writer.WriteString("resourceType", "Bundle");
            writer.WriteString("type", "transaction");
            writer.WriteStartArray("entry");
            foreach (var entry in entryWriters)
            {
                entry(writer);
            }
            writer.WriteEndArray();
            writer.WriteEndObject();
        }
        return Encoding.UTF8.GetString(stream.ToArray());
    }

    private static void WriteEntry(
        Utf8JsonWriter writer,
        PedigreeIndividual individual,
        string patientReference,
        FhirCoding coding)
    {
        writer.WriteStartObject();
        writer.WriteString("fullUrl", $"urn:uuid:{individual.Id}");

        writer.WriteStartObject("resource");
        writer.WriteString("resourceType", "FamilyMemberHistory");
        writer.WriteString("status", "completed");
        writer.WriteStartObject("patient");
        writer.WriteString("reference", patientReference);
        writer.WriteEndObject();
        writer.WriteString("name", individual.DisplayName);

        writer.WriteStartObject("relationship");
        writer.WriteStartArray("coding");
        writer.WriteStartObject();
        writer.WriteString("system", coding.System);
        writer.WriteString("code", coding.Code);
        writer.WriteString("display", coding.Display);
        writer.WriteEndObject();
        writer.WriteEndArray();
        writer.WriteEndObject();

        writer.WriteStartObject("sex");
        writer.WriteStartArray("coding");
        writer.WriteStartObject();
        writer.WriteString("system", AdminGenderSystem);
        writer.WriteString("code", BiologicalSexWire.Of(individual.BiologicalSex));
        writer.WriteEndObject();
        writer.WriteEndArray();
        writer.WriteEndObject();

        if (individual.BornOn is not null)
        {
            writer.WriteString("bornDate", individual.BornOn);
        }
        writer.WriteEndObject();

        writer.WriteStartObject("request");
        writer.WriteString("method", "POST");
        writer.WriteString("url", "FamilyMemberHistory");
        writer.WriteEndObject();

        writer.WriteEndObject();
    }
}

File: fhir-bridge/node/src/pedigreeToFhir.ts
/**
 * PedigreeDetail -> FHIR R5 transaction Bundle of FamilyMemberHistory.
 *
 * Pure transformation. Any relative whose relation cannot be mapped
 * produces a warning in `warnings`; the resource is skipped, never
 * fabricated.
 */

import type { FhirBundle, FhirBundleEntry, FhirFamilyMemberHistory } from './fhirTypes.js';
import type { PedigreeDetail, PedigreeIndividual } from './pedigreeDetail.js';
import { mapToFhir } from './relationMap.js';
import { relationsFromProband } from './probandRelations.js';

const ADMIN_GENDER_SYSTEM = 'http://hl7.org/fhir/administrative-gender';

export interface PedigreeMappingResult {
  readonly bundle: FhirBundle;
  readonly probandReference: string;
  readonly warnings: readonly string[];
}

export class MappingError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'MappingError';
  }
}

export function pedigreeToFhirBundle(detail: PedigreeDetail): PedigreeMappingResult {
  const relations = relationsFromProband(detail);
  if (relations === undefined) {
    throw new MappingError(
      `Pedigree ${detail.id} has no proband: set one before pushing to FHIR.`,
    );
  }

  const probandReference = `Patient/${relations.proband.id}`;
  const warnings: string[] = [];
  const entries: FhirBundleEntry[] = [];

  for (const view of relations.relatives) {
    const coding = mapToFhir(view.relativeType);
    if (coding === undefined) {
      warnings.push(
        `skipped ${view.individual.displayName}: no FHIR code for relative_type '${view.relativeType}'.`,
      );
      continue;
    }
    entries.push(buildEntry(view.individual, probandReference, coding.code, coding.display));
  }
  for (const unlabelled of relations.unlabelled) {
    warnings.push(
      `skipped ${unlabelled.displayName}: could not derive a supported relation to the proband.`,
    );
  }

  return {
    bundle: { resourceType: 'Bundle', type: 'transaction', entry: entries },
    probandReference,
    warnings,
  };
}

function buildEntry(
  individual: PedigreeIndividual,
  patientReference: string,
  fhirCode: string,
  fhirDisplay: string,
): FhirBundleEntry {
  const resource: FhirFamilyMemberHistory = {
    resourceType: 'FamilyMemberHistory',
    status: 'completed',
    patient: { reference: patientReference },
    name: individual.displayName,
    relationship: {
      coding: [
        {
          system: 'http://terminology.hl7.org/CodeSystem/v3-RoleCode',
          code: fhirCode,
          display: fhirDisplay,
        },
      ],
    },
    sex: {
      coding: [{ system: ADMIN_GENDER_SYSTEM, code: individual.biologicalSex }],
    },
    ...(individual.bornOn === undefined ? {} : { bornDate: individual.bornOn }),
  };
  return {
    fullUrl: `urn:uuid:${individual.id}`,
    resource,
    request: { method: 'POST', url: 'FamilyMemberHistory' },
  };
}
